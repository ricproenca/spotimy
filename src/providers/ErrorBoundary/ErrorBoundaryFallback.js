import Box from '@mui/material/Box';
import PropTypes from 'prop-types';
import React from 'react';

import { reloadPage } from '@Utils/router';

import ErrorBoundaryFallbackStyles from './ErrorBoundaryFallback.styles';

/**
 * Fallback UI for the Error boundary
 *
 * @param {object|string} error - error to show
 * @param {Function} resetError - callback to reset the Error
 */
const ErrorFallback = ({ error, resetError }) => {
  // eslint-disable-next-line no-console
  console.log('ErrorFallback -> error', error.message);

  const classes = ErrorBoundaryFallbackStyles();

  const showError = error.message ? error.message : error;

  return (
    <Box component='div' className={classes.root}>
      <h2 data-testid='title'>Ooops, something went wrong</h2>

      <pre className={classes.code} data-testid='error-message'>
        {showError}
      </pre>

      <button className={classes.button} onClick={resetError} data-testid='try-again-button'>
        Try again
      </button>

      <pre className={classes.stack}>{error.stack}</pre>
    </Box>
  );
};

ErrorFallback.propTypes = {
  /**
   * Error to show
   */
  error: PropTypes.oneOfType([PropTypes.object, PropTypes.string]),
  /**
   * callback to reset the Error
   */
  resetError: PropTypes.func
};

ErrorFallback.defaultProps = {
  error: '',
  resetError: reloadPage
};

export default ErrorFallback;
