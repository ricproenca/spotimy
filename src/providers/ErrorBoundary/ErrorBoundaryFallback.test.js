import '@testing-library/jest-dom';
import { fireEvent, render, screen } from '@testing-library/react';
import React from 'react';

import ErrorBoundaryFallback from './ErrorBoundaryFallback';

jest.mock('./ErrorBoundaryFallback.styles', () => () => ({ root: 'root', color: 'color' }));

const mockResetError = jest.fn();
const mockResetErrorCalls = mockResetError.mock.calls;

describe('ErrorBoundaryFallback', () => {
  describe('Testing rendering', () => {
    test('It matches snapshot', () => {
      // Act
      const { asFragment } = render(<ErrorBoundaryFallback />);
      // Assert
      expect(asFragment()).toMatchSnapshot();
    });

    test('It renders', () => {
      // Arrange
      const props = {
        error: { message: 'This is an error message' },
        resetError: mockResetError
      };

      // Act
      render(<ErrorBoundaryFallback {...props} />);

      const title = screen.getByTestId('title');
      const errorMessage = screen.getByTestId('error-message');
      const tryAgainButton = screen.getByTestId('try-again-button');

      // Assert
      expect(title).toBeInTheDocument();
      expect(errorMessage).toBeInTheDocument();
      expect(tryAgainButton).toBeInTheDocument();
    });
  });

  describe('Testing event handlers', () => {
    test('Call reset error', () => {
      // Arrange
      const props = {
        error: { message: 'This is an error message' },
        resetError: mockResetError
      };

      // Act
      render(<ErrorBoundaryFallback {...props} />);
      const tryAgainButton = screen.getByTestId('try-again-button');

      // Act
      fireEvent.click(tryAgainButton);

      // Assert
      expect(mockResetErrorCalls).toHaveLength(1);
    });
  });
});
