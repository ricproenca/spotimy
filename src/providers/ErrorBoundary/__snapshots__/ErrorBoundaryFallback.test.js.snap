// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundaryFallback Testing rendering It matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="root MuiBox-root css-0"
  >
    <h2
      data-testid="title"
    >
      Ooops, something went wrong
    </h2>
    <pre
      data-testid="error-message"
    />
    <button
      data-testid="try-again-button"
    >
      Try again
    </button>
    <pre />
  </div>
</DocumentFragment>
`;
